File: src/engine/careers.ts
import type { CharacterData, CareerData, TermSummary } from './types'
import { roll2d6, rollD6, rollVsTarget, characteristicModifier } from './dice'
import { loadCareer } from './dataLoader'
import { random } from './rng'

export function getStatDm(character: CharacterData, stat: string): number {
  const val = character.abilities[stat] || 7
  return characteristicModifier(val)
}

export function resolveQualification(character: CharacterData, career: CareerData): boolean {
  const qual = career.qualification
  const stat = qual.stat || 'end'
  let dm = (qual.dm || 0) + getStatDm(character, stat)
  // Apply -2 DM per previous career
  const uniqueCareers = new Set(character.careers.map(c => c.career_id))
  dm -= 2 * uniqueCareers.size

  const target = qual.target
  const { success } = rollVsTarget(target, dm)
  return success
}

export function gainSkill(character: CharacterData, skillId: string): void {
  if (skillId in character.abilities) {
    // Personal development stat increase
    const current = character.abilities[skillId] || 7
    if (current < 15) {
      character.abilities[skillId] = current + 1
    }
    return
  }

  const skillMax = character.derived['skill_maximum'] || 14
  const current = character.skills[skillId] || 0
  if (current < skillMax) {
    character.skills[skillId] = current + 1
  }
}

export function rollSkillTable(character: CharacterData, career: CareerData): string {
  const roll = rollD6()
  const skills = career.skills

  let table: string[]
  if (roll <= 2) {
    table = skills.personal
  } else if (roll <= 4) {
    table = skills.service
  } else if (roll === 5) {
    const edu = character.abilities['edu'] || 7
    if (edu >= 8) {
      table = skills.advanced || skills.service
    } else {
      table = skills.service
    }
  } else {
    table = skills.special || skills.service
  }

  return table[Math.floor(random() * table.length)]
}

export function resolveBasicTraining(character: CharacterData, career: CareerData): void {
  for (const skillId of career.skills.service) {
    if (!character.skills[skillId] && !(skillId in character.abilities)) {
      character.skills[skillId] = 0
    }
  }
}

export function resolveSkillRolls(character: CharacterData, career: CareerData, rolls: number = 1): void {
  for (let i = 0; i < rolls; i++) {
    const skillId = rollSkillTable(character, career)
    gainSkill(character, skillId)
  }
}

export function resolveSurvival(character: CharacterData, career: CareerData, term: number): { success: boolean; effect: number } {
  const surv = career.survival
  const stat = surv.stat || 'end'
  let dm = getStatDm(character, stat)

  const dmModifier = surv.dm_modifier || ''
  if (dmModifier === 'terms_served') {
    dm -= term - 1
  }

  const target = surv.target
  const { success, effect } = rollVsTarget(target, dm)
  return { success, effect }
}

export function resolveAdvancement(character: CharacterData, career: CareerData, effect: number): { advanced: boolean; newRank: number } {
  const adv = career.advancement
  const threshold = adv.effect_threshold || 4

  const currentCareerTerms = character.careers.filter(c => c.career_id === career.career_id)
  const currentRank = currentCareerTerms.length > 0 ? currentCareerTerms[currentCareerTerms.length - 1].rank : 0

  if (effect >= threshold && currentRank < career.ranks.length - 1) {
    const newRank = currentRank + 1
    const rankData = career.ranks[newRank]
    if (rankData.skill) {
      gainSkill(character, rankData.skill)
    }
    // Even ranks increase SOC by 1
    if (newRank % 2 === 0) {
      character.abilities['soc'] = (character.abilities['soc'] || 7) + 1
    }
    return { advanced: true, newRank }
  }
  return { advanced: false, newRank: currentRank }
}

export function resolveEvent(character: CharacterData, career: CareerData, currentRank: number): { description: string; newRank: number } {
  const roll = roll2d6()
  const events = career.events

  // Handle ranges like "3-4"
  let eventKey = String(roll)
  if (!events[eventKey]) {
    for (const key of Object.keys(events)) {
      if (key.includes('-')) {
        const [low, high] = key.split('-').map(Number)
        if (low <= roll && roll <= high) {
          eventKey = key
          break
        }
      }
    }
  }

  const event = events[eventKey] || { description: 'Uneventful term', effect: 'none' }
  const effect = event.effect
  let newRank = currentRank

  if (effect === 'skill_roll') {
    resolveSkillRolls(character, career, 1)
  } else if (effect === 'advancement_dm_plus_2') {
    resolveSkillRolls(character, career, 1)
  } else if (effect === 'automatic_advancement_plus_skill') {
    if (currentRank < career.ranks.length - 1) {
      newRank = currentRank + 1
      const rankData = career.ranks[newRank]
      if (rankData.skill) {
        gainSkill(character, rankData.skill)
      }
      character.abilities['soc'] = (character.abilities['soc'] || 7) + 1
    }
    resolveSkillRolls(character, career, 1)
  } else if (effect === 'aging_roll') {
    const stats = ['str', 'dex', 'end']
    const stat = stats[Math.floor(random() * stats.length)]
    character.abilities[stat] = Math.max(1, (character.abilities[stat] || 7) - 1)
  } else if (effect === 'mishap') {
    const mishapRoll = rollD6()
    const mishaps = career.mishaps
    const mishapDesc = mishaps[String(mishapRoll)] || 'Unknown mishap'
    return { description: `${event.description}: ${mishapDesc} (career ends)`, newRank: currentRank }
  }

  return { description: event.description, newRank }
}

export function resolveMishap(character: CharacterData, career: CareerData): string {
  const roll = rollD6()
  const mishaps = career.mishaps
  const desc = mishaps[String(roll)] || 'Unknown mishap'

  if (roll === 1) {
    const stats = ['str', 'dex', 'end']
    const stat = stats[Math.floor(random() * stats.length)]
    character.abilities[stat] = Math.max(1, (character.abilities[stat] || 7) - Math.floor(random() * 3) - 1)
  } else if (roll === 3) {
    character.abilities['soc'] = Math.max(2, (character.abilities['soc'] || 7) - 1)
  }

  return desc
}

export function resolveAging(character: CharacterData, term: number): boolean {
  if (term < 5) return true

  const endDm = getStatDm(character, 'end')
  const difficulty = term + 1
  const { success, effect } = rollVsTarget(difficulty, endDm)

  if (!success) {
    const stats = ['str', 'dex', 'end']
    const stat = stats[Math.floor(random() * stats.length)]
    character.abilities[stat] = Math.max(1, (character.abilities[stat] || 7) - 1)
    if (effect <= -6) {
      const stats2 = stats.filter(s => s !== stat)
      const stat2 = stats2[Math.floor(random() * stats2.length)]
      character.abilities[stat2] = Math.max(1, (character.abilities[stat2] || 7) - 1)
    }
  }
  return true
}

export function resolveReenlistment(career: CareerData): { canContinue: boolean; mandatory: boolean } {
  const target = career.reenlistment_target || 7
  const roll = roll2d6()
  if (roll === 12) {
    return { canContinue: true, mandatory: true }
  }
  return { canContinue: roll >= target, mandatory: false }
}

export function resolveTerm(character: CharacterData, careerId: string): TermSummary {
  const career = loadCareer(careerId)
  const termNum = character.careers.filter(c => c.career_id === careerId).length + 1

  const isFirstTermInCareer = termNum === 1
  const isFirstCareer = character.careers.length === 0

  const termSummary: TermSummary = {
    career_id: careerId,
    term: termNum,
    start_age: character.age,
    survived: true,
    mishap: null,
    advanced: false,
    rank: 0,
    event: '',
    skills_gained: [],
  }

  if (isFirstTermInCareer) {
    if (!isFirstCareer) {
      const qualified = resolveQualification(character, career)
      if (!qualified) {
        termSummary.survived = false
        termSummary.mishap = 'Failed qualification'
        return termSummary
      }
    }

    resolveBasicTraining(character, career)
  }

  // Skill rolls
  let skillRolls = 1
  if (isFirstTermInCareer) {
    skillRolls = 2
  }

  // Survival
  const { success, effect } = resolveSurvival(character, career, termNum)
  if (!success) {
    const mishapDesc = resolveMishap(character, career)
    termSummary.survived = false
    termSummary.mishap = mishapDesc
    character.age += 4
    character.total_terms += 1
    return termSummary
  }

  // Advancement
  const { advanced, newRank } = resolveAdvancement(character, career, effect)
  termSummary.advanced = advanced
  termSummary.rank = newRank
  if (advanced) {
    skillRolls += 1
  }

  // Apply skill rolls
  for (let i = 0; i < skillRolls; i++) {
    const skillId = rollSkillTable(character, career)
    gainSkill(character, skillId)
    termSummary.skills_gained.push(skillId)
  }

  // Event
  const { description: eventDesc, newRank: eventRank } = resolveEvent(character, career, newRank)
  termSummary.event = eventDesc
  if (eventRank > termSummary.rank) {
    termSummary.rank = eventRank
    termSummary.advanced = true
  }

  // Aging
  const totalTerms = character.total_terms + 1
  resolveAging(character, totalTerms)

  // Update character state
  character.age += 4
  character.total_terms = totalTerms
  character.careers.push(termSummary)

  return termSummary
}

export function musterOut(character: CharacterData): { cashTotal: number; benefits: string[] } {
  const benefits: { cashRolls: number[]; cashTotal: number; benefitRolls: number[]; benefitsGained: string[] } = {
    cashRolls: [],
    cashTotal: 0,
    benefitRolls: [],
    benefitsGained: [],
  }

  const careerSummary: Record<string, { terms: number; rank: number }> = {}
  for (const term of character.careers) {
    const cid = term.career_id
    if (!careerSummary[cid]) {
      careerSummary[cid] = { terms: 0, rank: 0 }
    }
    careerSummary[cid].terms += 1
    careerSummary[cid].rank = Math.max(careerSummary[cid].rank, term.rank)
  }

  for (const [cid, summary] of Object.entries(careerSummary)) {
    const career = loadCareer(cid)
    const rank = summary.rank
    const terms = summary.terms

    // Cash rolls
    let cashDm = rank
    if (character.skills['gambling']) {
      cashDm += Math.floor(character.skills['gambling'] / 2)
    }

    const maxCashRolls = Math.min(3 - benefits.cashRolls.length, terms)
    for (let i = 0; i < maxCashRolls; i++) {
      let roll = rollD6() + cashDm
      roll = Math.max(1, Math.min(roll, career.mustering.cash.length))
      const cashAmount = career.mustering.cash[roll - 1]
      benefits.cashRolls.push(roll)
      benefits.cashTotal += cashAmount
    }

    // Benefit rolls
    const benefitDm = rank
    for (let i = 0; i < terms; i++) {
      let roll = rollD6() + benefitDm
      roll = Math.max(1, Math.min(roll, career.mustering.benefits.length))
      const benefit = career.mustering.benefits[roll - 1]
      benefits.benefitRolls.push(roll)
      benefits.benefitsGained.push(benefit)
    }
  }

  if (benefits.cashTotal > 0) {
    character.assets.push({ type: 'cash', amount: benefits.cashTotal })
  }

  const careerIds = Object.keys(careerSummary)
  for (let i = 0; i < benefits.benefitsGained.length; i++) {
    const benefit = benefits.benefitsGained[i]
    const sourceCareer = careerIds[i % careerIds.length] || 'unknown'
    if (benefit === 'weapon' || benefit === 'armor') {
      character.equipment.push({ id: benefit, condition: 'good' })
    } else {
      character.assets.push({ type: benefit, description: `Gained from ${sourceCareer}` })
    }
  }

  return { cashTotal: benefits.cashTotal, benefits: benefits.benefitsGained }
}

File: python/src/careers.py
"""Career term resolution and mustering out."""

import random
from typing import Dict, List, Optional

from dice import roll_2d6, roll_d6, roll_vs_target, characteristic_modifier
from character import Character, load_json, data_path


def load_career(career_id: str) -> dict:
    """Load a single career JSON file."""
    return load_json(data_path("careers", f"{career_id}.json"))


def get_stat_dm(character: Character, stat: str) -> int:
    """Get the DM for a characteristic."""
    val = character.abilities.get(stat, 7)
    return characteristic_modifier(val)


def resolve_qualification(character: Character, career: dict) -> bool:
    """Attempt to qualify for a career. Returns True on success."""
    qual = career["qualification"]
    stat = qual.get("stat", "end")
    dm = qual.get("dm", 0) + get_stat_dm(character, stat)
    # Apply -2 DM per previous career
    dm -= 2 * len({c["career_id"] for c in character.careers})

    target = qual["target"]
    roll, total, success, effect = roll_vs_target(target, dm)
    return success


def gain_skill(character: Character, skill_id: str) -> None:
    """Increase a skill by one level, respecting skill maximum."""
    if skill_id in character.abilities:
        # Personal development stat increase
        stat = skill_id
        current = character.abilities.get(stat, 7)
        if current < 15:
            character.abilities[stat] = current + 1
        return

    skill_max = character.derived.get("skill_maximum", 14)
    current = character.skills.get(skill_id, 0)
    if current < skill_max:
        character.skills[skill_id] = current + 1


def roll_skill_table(character: Character, career: dict) -> str:
    """Roll on a career's skill table and return the skill gained."""
    roll = roll_d6()
    skills = career["skills"]

    if roll <= 2:
        table = skills["personal"]
    elif roll <= 4:
        table = skills["service"]
    elif roll == 5:
        edu = character.abilities.get("edu", 7)
        if edu >= 8:
            table = skills.get("advanced", skills["service"])
        else:
            table = skills["service"]
    else:
        # Roll 6: special or fallback to service
        table = skills.get("special", skills["service"])

    return random.choice(table)


def resolve_basic_training(character: Character, career: dict) -> None:
    """Gain all service skills at Level 0."""
    for skill_id in career["skills"]["service"]:
        if skill_id not in character.skills and skill_id not in character.abilities:
            character.skills[skill_id] = 0


def resolve_skill_rolls(character: Character, career: dict, rolls: int = 1) -> None:
    """Roll for skills the given number of times."""
    for _ in range(rolls):
        skill_id = roll_skill_table(character, career)
        gain_skill(character, skill_id)


def resolve_survival(character: Character, career: dict, term: int) -> tuple:
    """Roll survival. Returns (success, effect)."""
    surv = career["survival"]
    stat = surv.get("stat", "end")
    dm = get_stat_dm(character, stat)

    # terms_served penalty if applicable
    dm_modifier = surv.get("dm_modifier", "")
    if dm_modifier == "terms_served":
        dm -= term - 1  # penalty increases each term

    target = surv["target"]
    roll, total, success, effect = roll_vs_target(target, dm)
    return success, effect


def resolve_advancement(character: Character, career: dict, effect: int) -> tuple:
    """Check advancement based on survival effect. Returns (advanced, new_rank)."""
    adv = career["advancement"]
    threshold = adv.get("effect_threshold", 4)

    current_career_terms = [c for c in character.careers if c["career_id"] == career["career_id"]]
    current_rank = current_career_terms[-1]["rank"] if current_career_terms else 0

    if effect >= threshold and current_rank < len(career["ranks"]) - 1:
        new_rank = current_rank + 1
        rank_data = career["ranks"][new_rank]
        if rank_data.get("skill"):
            gain_skill(character, rank_data["skill"])
        # Even ranks increase SOC by 1
        if new_rank % 2 == 0:
            character.abilities["soc"] = character.abilities.get("soc", 7) + 1
        return True, new_rank
    return False, current_rank


def resolve_event(character: Character, career: dict, current_rank: int = 0) -> tuple:
    """Roll a career event and apply effects. Returns (description, new_rank)."""
    roll = roll_2d6()
    event_key = str(roll)
    events = career.get("events", {})

    # Handle ranges like "3-4"
    if event_key not in events:
        for key in events:
            if "-" in key:
                low, high = key.split("-")
                if int(low) <= roll <= int(high):
                    event_key = key
                    break

    event = events.get(event_key, {"description": "Uneventful term", "effect": "none"})
    effect = event.get("effect", "none")
    new_rank = current_rank

    if effect == "skill_roll":
        resolve_skill_rolls(character, career, 1)
    elif effect == "advancement_dm_plus_2":
        # For next term only; simplified: just grant an extra skill
        resolve_skill_rolls(character, career, 1)
    elif effect == "automatic_advancement_plus_skill":
        # Auto-advance if possible
        if current_rank < len(career["ranks"]) - 1:
            new_rank = current_rank + 1
            rank_data = career["ranks"][new_rank]
            if rank_data.get("skill"):
                gain_skill(character, rank_data["skill"])
            character.abilities["soc"] = character.abilities.get("soc", 7) + 1
        resolve_skill_rolls(character, career, 1)
    elif effect == "aging_roll":
        # Simplified: lose 1 from STR, DEX, or END
        stat = random.choice(["str", "dex", "end"])
        character.abilities[stat] = max(1, character.abilities.get(stat, 7) - 1)
    elif effect == "mishap":
        # Roll mishap and eject from career
        mishap_roll = roll_d6()
        mishaps = career.get("mishaps", {})
        mishap_desc = mishaps.get(str(mishap_roll), "Unknown mishap")
        return f"{event['description']}: {mishap_desc} (career ends)", current_rank

    return event["description"], new_rank


def resolve_mishap(character: Character, career: dict) -> str:
    """Roll on mishap table. Returns description."""
    roll = roll_d6()
    mishaps = career.get("mishaps", {})
    desc = mishaps.get(str(roll), "Unknown mishap")

    if roll == 1:
        # Severely injured
        stat = random.choice(["str", "dex", "end"])
        character.abilities[stat] = max(1, character.abilities.get(stat, 7) - random.randint(1, 3))
    elif roll == 3:
        character.abilities["soc"] = max(2, character.abilities.get("soc", 7) - 1)

    return desc


def resolve_aging(character: Character, term: int) -> bool:
    """Resolve aging for term 5+ (age 34+). Returns True if survived aging."""
    if term < 5:
        return True

    end_dm = get_stat_dm(character, "end")
    difficulty = term + 1  # Terms served + 1
    roll, total, success, effect = roll_vs_target(difficulty, end_dm)

    if not success:
        # Lose 1 point from STR, DEX, or END
        stat = random.choice(["str", "dex", "end"])
        character.abilities[stat] = max(1, character.abilities.get(stat, 7) - 1)
        if effect <= -6:
            # Critical failure: lose from two abilities
            stat2 = random.choice([s for s in ["str", "dex", "end"] if s != stat])
            character.abilities[stat2] = max(1, character.abilities.get(stat2, 7) - 1)
    return True


def resolve_reenlistment(career: dict) -> tuple:
    """Roll re-enlistment. Returns (can_continue, mandatory)."""
    # Default re-enlistment target is 7 for most careers
    target = career.get("reenlistment_target", 7)
    roll = roll_2d6()
    if roll == 12:
        return True, True  # Mandatory continuation
    return roll >= target, False


def resolve_term(character: Character, career_id: str) -> dict:
    """Resolve a single career term. Returns term summary."""
    career = load_career(career_id)
    term_num = len([c for c in character.careers if c["career_id"] == career_id]) + 1

    # Check if this is first term in this career
    is_first_term_in_career = term_num == 1
    is_first_career = len(character.careers) == 0

    term_summary = {
        "career_id": career_id,
        "term": term_num,
        "start_age": character.age,
        "survived": True,
        "mishap": None,
        "advanced": False,
        "rank": 0,
        "event": "",
        "skills_gained": [],
    }

    if is_first_term_in_career:
        # Qualification (skip if first career and auto-qualify)
        if not is_first_career:
            qualified = resolve_qualification(character, career)
            if not qualified:
                term_summary["survived"] = False
                term_summary["mishap"] = "Failed qualification"
                return term_summary

        # Basic training
        resolve_basic_training(character, career)

    # Skill rolls: 1 per term, +1 if first term and no advancement
    skill_rolls = 1
    if is_first_term_in_career:
        skill_rolls = 2  # Simplified: 2 rolls first term

    # Survival
    survived, effect = resolve_survival(character, career, term_num)
    if not survived:
        mishap_desc = resolve_mishap(character, career)
        term_summary["survived"] = False
        term_summary["mishap"] = mishap_desc
        character.age += 4
        character.total_terms += 1
        character.careers.append(term_summary)
        return term_summary

    # Advancement (Mneme: Effect 4+ on survival)
    advanced, new_rank = resolve_advancement(character, career, effect)
    term_summary["advanced"] = advanced
    term_summary["rank"] = new_rank
    if advanced:
        skill_rolls += 1

    # Apply skill rolls
    for _ in range(skill_rolls):
        skill_id = roll_skill_table(character, career)
        gain_skill(character, skill_id)
        term_summary["skills_gained"].append(skill_id)

    # Event
    event_desc, event_rank = resolve_event(character, career, new_rank)
    term_summary["event"] = event_desc

    # Update rank if event caused automatic advancement
    if event_rank > term_summary["rank"]:
        term_summary["rank"] = event_rank
        term_summary["advanced"] = True

    # Aging
    total_terms = character.total_terms + 1
    resolve_aging(character, total_terms)

    # Update character state
    character.age += 4
    character.total_terms = total_terms
    character.careers.append(term_summary)

    return term_summary


def muster_out(character: Character) -> dict:
    """Resolve mustering out benefits for all completed careers."""
    benefits = {
        "cash_rolls": [],
        "cash_total": 0,
        "benefit_rolls": [],
        "benefits_gained": [],
    }

    # Group careers by career_id to find best term count per career
    career_summary: Dict[str, dict] = {}
    for term in character.careers:
        cid = term["career_id"]
        if cid not in career_summary:
            career_summary[cid] = {"terms": 0, "rank": 0}
        career_summary[cid]["terms"] += 1
        career_summary[cid]["rank"] = max(career_summary[cid]["rank"], term["rank"])

    for cid, summary in career_summary.items():
        career = load_career(cid)
        rank = summary["rank"]
        terms = summary["terms"]

        # Cash rolls: up to 3 across all careers
        cash_dm = rank
        if "gambling" in character.skills:
            cash_dm += character.skills["gambling"] // 2

        max_cash_rolls = min(3 - len(benefits["cash_rolls"]), terms)
        for _ in range(max_cash_rolls):
            roll = roll_d6() + cash_dm
            roll = max(1, min(roll, len(career["mustering"]["cash"])))
            cash_amount = career["mustering"]["cash"][roll - 1]
            benefits["cash_rolls"].append(roll)
            benefits["cash_total"] += cash_amount

        # Benefit rolls: 1 per term
        benefit_dm = rank
        for _ in range(terms):
            roll = roll_d6() + benefit_dm
            roll = max(1, min(roll, len(career["mustering"]["benefits"])))
            benefit = career["mustering"]["benefits"][roll - 1]
            benefits["benefit_rolls"].append(roll)
            benefits["benefits_gained"].append(benefit)

    # Add cash to assets
    if benefits["cash_total"] > 0:
        character.assets.append({"type": "cash", "amount": benefits["cash_total"]})

    # Add material benefits as assets/equipment
    career_ids = list(career_summary.keys())
    for idx, benefit in enumerate(benefits["benefits_gained"]):
        source_career = career_ids[idx % len(career_ids)] if career_ids else "unknown"
        if benefit in ["weapon", "armor"]:
            character.equipment.append({"id": benefit, "condition": "good"})
        else:
            character.assets.append({"type": benefit, "description": f"Gained from {source_career}"})

    character.derived.update(benefits)
    return benefits
